app/src/rgb_underglow.c

```diff
@@ -468,17 +468,28 @@ int zmk_rgb_underglow_change_spd(int direction) {
 
 #if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_AUTO_OFF_IDLE) ||                                          \
     IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_AUTO_OFF_USB)
-static int rgb_underglow_auto_state(bool *prev_state, bool new_state) {
-    if (state.on == new_state) {
+struct rgb_underglow_sleep_state {
+    bool sleeping;
+    bool pre_sleep;
+};
+
+static int rgb_underglow_auto_state(bool is_wakeup_event) {
+    static struct rgb_underglow_sleep_state sleep_state = {sleeping : false, pre_sleep : false};
+
+    // wake up event while awake, or sleep event while sleeping -> no-op
+    if (!is_wakeup_event == sleep_state.sleeping) {
         return 0;
     }
-    if (new_state) {
-        state.on = *prev_state;
-        *prev_state = false;
-        return zmk_rgb_underglow_on();
+    sleep_state.sleeping = !is_wakeup_event;
+
+    if (is_wakeup_event) {
+        if (sleep_state.pre_sleep) {
+            return zmk_rgb_underglow_on();
+        } else {
+            return zmk_rgb_underglow_off();
```

Comment:
we can probably remove this line / return 0 instead here

Reply:
If my understanding was correct a few days ago while reading zephyr's docs about settings system, it checks if data is different before writing.

And i think having an explicit `off()` instrad of `return 0` is better because is much more obvious and readable on a quick glance, even though it is equivalent logic-wise.